q2_feature_table/_summarize/summarize_assets/src/chart.js

```diff
@@ -0,0 +1,165 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) 2016-2017, QIIME 2 development team.
+//
+// Distributed under the terms of the Modified BSD License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+// ----------------------------------------------------------------------------
+
+import {
+  select,
+  selectAll,
+  histogram,
+  scaleBand,
+  scaleLinear,
+  max,
+  axisBottom,
+  axisLeft,
+} from 'd3';
+
+let DROPPED = [];
+
+const buildBins = (data, x) => (histogram()(data.map(x)));
+
+
+const updateChart = (metadata, props, xScale, yScale) => {
+  const option = select('select').node().value;
+  const validKeys = Object.keys(metadata[option]).filter(key => !DROPPED.includes(key));
+  const data = validKeys.map(key => metadata[option][key]);
+
+  const bins = buildBins(data, xScale);
+
+  selectAll('.overlay-group')
+    .data(bins)
+  .select('.overlay')
+    .attr('y', d => yScale(d.length))
+    .attr('height', d => props.height - yScale(d.length));
+};
+
+export const addSampleMetadata = (sampleID) => {
+  if (DROPPED.includes(sampleID)) {
+    DROPPED = DROPPED.filter(i => i !== sampleID);
+  }
+};
+
+export const dropSampleMetadata = (sampleID) => {
+  if (!DROPPED.includes(sampleID)) {
+    DROPPED.push(sampleID);
+  }
+};
+
+const buildChart = (svg, metadata, props) => {
+  const option = select('select').node().value;
+
+  const data = Object.keys(metadata[option]).map(key => metadata[option][key]);
+  const setArray = [...new Set(data)];
+
+  svg.selectAll('g').remove();
+  svg.selectAll('rect').remove();
+
+  const g = svg
+    .append('g')
+      .attr('transform', `translate(${props.margin.left}, ${props.margin.top})`);
+
+  const x = scaleBand()
+    .domain(setArray)
+    .rangeRound([0, props.width - props.margin.left - props.margin.right]);
+
+  const bins = buildBins(data, x);
+
+  const y = scaleLinear()
+    .domain([0, max(bins, d => d.length)])
+    .range([props.height, 0]);
+
+  const bar = g.selectAll('.bar')
+    .data(bins)
+    .enter()
+      .append('g')
+        .attr('class', 'bar');
+        // .attr('transform', d => `translate(${d.x0}, ${y(d.length)})`);
+
+  const xAxis = axisBottom(x);
+  const yAxis = axisLeft(y);
+
+  bar.append('rect')
+    .attr('fill', 'black')
+    .attr('opacity', 0.15)
+    .attr('x', d => d.x0)
+    .attr('y', d => y(d.length))
+    .attr('width', x.bandwidth() - 1)
+    .attr('height', d => props.height - y(d.length))
+  .select(function backToBar() { return this.parentNode; })
```

Comment:
oh wow, do the callback's not provide the node as an argument? This might warrant a comment, as the correct behavior depends on how `this` is bound (requiring the old-style function)